### ISSUE TITLE
Align which-key and mappings.lua configuration

### ISSUE THREAD

**Issue by jayp0521** [AstroNvim/AstroNvim#781]
**Is your feature request related to a problem? Please describe.**
Hello,

I was trying to add some mappings to `mappings.lua` in this format:

```lua
		["<leader>"] = {
			x = {
				e = {
					function()
						require("dapui").eval()
					end,
					desc = "Evaluate Line",
				},
			},
		},
```

After much digging, It seems like this is supported in only `which-key` registration.

**Describe the solution you'd like**

Keep the same syntax support for both `mappings.lua` and `which-key` or only support one style. Two is just confusing and probably more maintenance. 

**Additional context**

**Comment by jayp0521** [comment]
If this is supposed to work, then this is a bug report

**Comment by osamuaoki** [comment]
Hi,

I think OP is talking about split configuration case.

I set up custom key assignment using user/init.lua file as follows:

`["which-key"]` and `mappings` are used for different purpose.
* `["which-key"]` is for the opening screen entries.
* `mappings` is for the next level menu data and key bindings.

The following will override package default bindings and add "+Buffer" menu.

```
 -- Modify which-key registration
  ["which-key"] = {
    -- Add bindings which show up as group
    register_mappings = {
      -- first key is the mode, n == normal mode
      n = {
        -- second key is the prefix, <leader> prefixes
        ["<leader>"] = {
          ["b"] = { name = "Buffer" },
        },
      },
    },
  },

  mappings = {
    -- first key is the mode
    n = {
      -- second key is the lefthand side of the map
      -- Terminal
      -- Install ncdu package
      -- Disk siz w/o snapshot btrfs subvolumes and .git
      ["<leader>tu"] = { function() astronvim.toggle_term_cmd "ncdu -x --exclude .git" end, desc = "Term for NCDU w/o junks" },
      -- Install mc package
      ["<leader>tm"] = { function() astronvim.toggle_term_cmd "mc . ~" end, desc = "Term for MC" },
      -- Install libreply-perl + libterm-readline-gnu-perl packages
      ["<leader>tp"] = { function() astronvim.toggle_term_cmd "reply" end, desc = "Term for perl5 REPL" },
      -- BUffer
      ["<leader>bb"] = { "<cmd>tabnew<cr>", desc = "New tab" },
      ["<leader>bc"] = { "<cmd>BufferLinePickClose<cr>", desc = "Pick to close" },
      ["<leader>bd"] = { "<cmd>BufferLineSortByDirectory<cr>", desc = "Sort by directory" },
      ["<leader>bD"] = { "<cmd>BufferLineSortByRelativeDirectory<cr>", desc = "Sort by rel. dir." },
      ["<leader>be"] = { "<cmd>BufferLineSortByExtension<cr>", desc = "Sort by extention" },
      ["<leader>bf"] = { "<cmd>Telescope buffers<cr>", desc = "Find" },
      ["<leader>bj"] = { "<cmd>BufferLinePick<cr>", desc = "Pick to jump" },
      ["<leader>bl"] = { "<cmd>BufferLineCloseLeft<cr>", desc = "Close left-side" },
      ["<leader>bn"] = { "<cmd>BufferLineCycleNext<cr>", desc = "Next" },
      ["<leader>bp"] = { "<cmd>BufferLineCyclePrev<cr>", desc = "Previous" },
      ["<leader>br"] = { "<cmd>BufferLineCloseRight<cr>", desc = "Close right-side" },
      ["<leader>bt"] = { "<cmd>BufferLineSortByTabs<cr>", desc = "Sort by tabs" },
      -- quick save
      -- ["<C-s>"] = { ":w!<cr>", desc = "Save File" },  -- change description but the same command
    },
    t = {
      -- setting a mapping to false will disable it
      -- ["<esc>"] = false,
    },
  },
```                                                                                                                            
Please pay attention to `name` and `desc`.

These are used for different purpose.  So you should follow similar with split setting.

I hope this might help you.

**Comment by osamuaoki** [comment]
I looked into this problem.  

* I think key-mappings should be declared in `mappings`.
* Its `desc` entry is stored by vim.map.set() as a part of table of opts.
* Then, which-key reads vim mapping data to generate most menu entry.
* The which-key menu entry to next level still needs to be declared in `["which-key"]`.
* Configuration parsing order should be in sync with the entry in  user_examples/init.lua to reduce confusion.

In light of the above, I propose to update pertinent part of init.lua as follows (reversed order and extra comment):

```
  mappings = {
    -- first key is the mode
    -- desc setting is stored by vim.keymap.set() as a part of opts table in vim lua module
    n = {
      -- second key is the lefthand side of the map
      -- Terminal
      -- Install ncdu package: (Disk size w/o snapshot btrfs subvolumes and .git)
      ["<leader>tu"] = { function() astronvim.toggle_term_cmd "ncdu -x --exclude .git" end, desc = "Term for NCDU w/o junks" },
      -- Install mc package:
      ["<leader>tm"] = { function() astronvim.toggle_term_cmd "mc . ~" end, desc = "Term for MC" },
      -- Install libreply-perl + libterm-readline-gnu-perl packages:
      ["<leader>tp"] = { function() astronvim.toggle_term_cmd "reply" end, desc = "Term for perl5 REPL" },
      -- BUffer
      ["<leader>bb"] = { "<cmd>tabnew<cr>", desc = "New tab" },
      ["<leader>bc"] = { "<cmd>BufferLinePickClose<cr>", desc = "Pick to close" },
      ["<leader>bd"] = { "<cmd>BufferLineSortByDirectory<cr>", desc = "Sort by directory" },
      ["<leader>bD"] = { "<cmd>BufferLineSortByRelativeDirectory<cr>", desc = "Sort by rel. dir." },
      ["<leader>be"] = { "<cmd>BufferLineSortByExtension<cr>", desc = "Sort by extention" },
      ["<leader>bf"] = { "<cmd>Telescope buffers<cr>", desc = "Find" },
      ["<leader>bj"] = { "<cmd>BufferLinePick<cr>", desc = "Pick to jump" },
      ["<leader>bl"] = { "<cmd>BufferLineCloseLeft<cr>", desc = "Close left-side" },
      ["<leader>bn"] = { "<cmd>BufferLineCycleNext<cr>", desc = "Next" },
      ["<leader>bp"] = { "<cmd>BufferLineCyclePrev<cr>", desc = "Previous" },
      ["<leader>br"] = { "<cmd>BufferLineCloseRight<cr>", desc = "Close right-side" },
      ["<leader>bt"] = { "<cmd>BufferLineSortByTabs<cr>", desc = "Sort by tabs" },
      -- quick save
      -- ["<C-s>"] = { ":w!<cr>", desc = "Save File" },  -- change description but the same command
    },
    t = {
      -- setting a mapping to false will disable it
      -- ["<esc>"] = false,
    },
  },

  -- Modify which-key registration (this is parsed after mapping)
  ["which-key"] = {
    -- Add bindings which show up as group name
    -- 
    -- Don't create settings for key mapping here.  Such mapping becomes usable
    -- only via slow which-key interface (not usable via native fast vim key
    -- mapping)
    --
    -- which-key will pick up defined vim key mappings, if defined with desc.
    register_mappings = {
      -- first key is the mode, n == normal mode
      n = {
        -- second key is the prefix, <leader> prefixes
        ["<leader>"] = {
          -- third key is the key to bring up next level and its displayed name in which-key top level
          ["b"] = { name = "Buffer" },
        },
      },
    },
  },
```

**Comment by osamuaoki** [comment]
> If this is supposed to work, then this is a bug report

I think this is a user configuration issue and slight lack of documentation.

First user issue: don't try to define key mapping in `['which-key"]` table.  Use `"mappings"` table.  (See above example)

Second, documentation: Please consider updating the example and documentation.

**Comment by mehalter** [comment]
@osamuaoki thank you for providing such a good description! They are correct that these two setting locations achieve different goals and are wrapping completely different APIs which is why their interfaces are different. I do agree that the recommended documentation changes provided really clear up the user documentation :)

**Comment by jayp0521** [comment]
> > If this is supposed to work, then this is a bug report
> 
> I think this is a user configuration issue and slight lack of documentation.
> 
> First user issue: don't try to define key mapping in `['which-key"]` table. Use `"mappings"` table. (See above example)
> 
> Second, documentation: Please consider updating the example and documentation.

> First user issue: don't try to define key mapping in ['which-key"] table. Use "mappings" table. (See above example)

The problem with this is that it is harder to maintain like in `which-key` I can do 

```
		["<leader>"] = {
			x = {
				e = {
					function()
						require("dapui").eval()
					end,
					desc = "Evaluate Line",
				},
			},
		},
```

In mappings.lua, I have to do. It would be nice to have the above functionality in `mappings.lua`

```
["<leader>xe"] = {
	          function()
		          require("dapui").eval()
	          end,
	          desc = "Evaluate Line",
},
```

**Comment by mehalter** [comment]
@jayp0521 in accordance to the contribution guidelines and direction of AstroNvim we are trying to minimize the amount of abstraction away from how things are done in a standard Neovim config. We have implemented a table that reflects how these API calls are structured in their own rights which 1. Makes it easy to convert existing Lua configs to AstroNvim in an intelligent way and 2. Makes it easy to go from an AstroNvim config to a vanilla Neovim config and understand exactly what each like is contributing to. Because of this we will not be changing the mappings API to match what which-key has done. Naturally which-key mappings have a nested menu like structure so this API structure makes complete sense. When defining just a list of mappings this is not necessarily the case semantically. You can feel free to implement your mappings such as leader mappings in the which-key table. This is actually how I do it in my own config to utilize this nested map like structure. There is no reason you *have* to use the mappings table. You can also always decide to implement your own mappings API in the `polish` function and make something to perfectly fit your needs :)

**Comment by osamuaoki** [comment]
> @osamuaoki thank you for providing such a good description! They are correct that these two setting locations achieve different goals and are wrapping completely different APIs which is why their interfaces are different. I do agree that the recommended documentation changes provided really clear up the user documentation :)

>@mehalter
Without your help, I couldn't learn so quickly.  Thanks.

Biggest learning curve for this well thought out AstroNvim is to understand what does this `user_plugin_opts` shim function do ?
Although we need to read the source eventually, I made a try to explain overall picture in plain words.  When you have time, please look at what I did as  AstroNvim/astronvim.github.io#13

**Comment by jayp0521** [comment]
> contribution guidelines and direction of AstroNvim

is this documented somewhere? don't mean to sound rude, would like to read it before creating issues. did not see anything [here](https://github.com/AstroNvim/AstroNvim/blob/main/.github/CONTRIBUTING.md)

> 1. Makes it easy to convert existing Lua configs to AstroNvim in an intelligent way and 2. Makes it easy to go from an AstroNvim config to a vanilla Neovim config and understand exactly what each like is contributing to.

unsure how based on the above, it was determined that the current (new) `mapping.lua` structure aligns with those points. Does 1 mean existing config of users in AstroNvim. I did not mind the conversion, but I wouldn't say it was easy to convert. https://github.com/jayp0521/dotfiles/commit/cd3996d58d13e4280b6f5f7a3b62eabbaccfbf75

For point 2, I mean no matter what the structure is, you can just copy and paste the [function](https://github.com/AstroNvim/AstroNvim/blob/main/lua/core/utils/init.lua#L218-L233) and get the same functionality. Do you mean the lines of code would be shorter in this format? Not sure if people logically understand that `index = 1` should be the `cmd` and the rest of the table should be the mapping `options`. I did not until I read the `set_mappings` function.

Ultimately, the decision is up to you. However, from what I can see is that there is the existing lua api `vim.keymap.set`, the `which-key` structure/api, and now there is another way of configuring keymappings (a wrapper around `vim.keymap.set`). Usually there is some shortcoming of the existing methods like too many lines of code, confusing, etc. before inventing a new way of doing something like diverging from an existing option like the `which-key` structure. Thanks for the time
